### ISSUE TITLE
[Feature Request] Article search functionality - via text and/or topics

### ISSUE THREAD

**Issue by kera-cudmore** [codu-code/codu#393]
## Context

As there are quite a number of articles on the site now 😊 it would be great to have search functionality, for text and/or tags

## Expected Behaviour

- Search using  text input **'React'** would return all articles that contain the word React in the title and/or excerpt
- Search via topics would return all articles with that topic tag (would be great to see all the different topics there are)

## What is the current behaviour?

Currently can only search topics that are hard-coded into the home page, or by clicking on a topic on a published article

**Comment by NiallJoeMaher** [comment]
Great idea @kera-cudmore! Let me look into it and come up with a plan unless someone else has ideas to add to this thread.

**Comment by EvanBurbidge** [comment]
Could be quite easily done there is already a small level of filtering with the "recommended topics".

Could add a search bar across the top that will provide filters such as "text search" for article names, and also a dropdown containing authors to filter by author.

A second step could be to add advanced filters to allow filtering between date ranges to get articles published between certain dates etc.

Here's a quick sketch of where those filters could fit

<img width="871" alt="Screenshot 2023-08-29 at 11 34 34" src="https://github.com/codu-code/codu/assets/10980413/875153bc-c636-461b-8845-d10d5d7a8c01">

**Comment by John-Paul-Larkin** [comment]
I gave this a go. [View](https://youtu.be/3zhH449joUk) the result.

I am still trying to figure things out, so you will have to forgive my poor commit history. I originally had a number of separate commits, but ended up doing a reset and combining them into a single one. 

The search part was pretty simple, but I used shadcn to implement the date picker, which added 8 packages, including radix ui. It added utility classes for the light/dark mode to the styles/globals.css. Some of these are not strictly necessary, but I left them in, as I was not sure how the light/dark mode will be implemented. It also added a number of custom themes to the tailwind.config.js. I don't know if these will cause conflicts down the line?

This is a learning exercise for me, so I have no issue with scrapping the lot, or going back to make any changes. Feedback, or resources to help me through the contribution process would be greatly appreciated. 

PS Assuming the bug fix(random bookmarks) from issus 416, has been deployed, it is not working on codu.co
